lib/elastic_apm/transport/base.rb

```diff
@@ -8,90 +8,156 @@
 
 module ElasticAPM
   module Transport
+    # rubocop:disable Metrics/ClassLength
     # @api private
     class Base
       include Logging
 
+      WATCHER_EXECUTION_INTERVAL = 5
+      WATCHER_TIMEOUT_INTERVAL = 4
+      WORKER_JOIN_TIMEOUT = 5
+
       def initialize(config)
         @config = config
         @queue = SizedQueue.new(config.api_buffer_size)
-        @pool = Concurrent::FixedThreadPool.new(config.pool_size)
-        @workers = []
 
         @serializers = Serializers.new(config)
         @filters = Filters.new(config)
+
+        @stopped = Concurrent::AtomicBoolean.new
+        @workers = Array.new(config.pool_size)
+
+        @watcher_mutex = Mutex.new
+        @worker_mutex = Mutex.new
       end
 
-      attr_reader :config, :queue, :workers, :filters
+      attr_reader :config, :queue, :filters, :workers, :watcher, :stopped
 
       def start
+        debug '%s: Starting Transport', pid_str
+
+        ensure_watcher_running
+        ensure_worker_count
       end
 
       def stop
+        debug '%s: Stopping Transport', pid_str
+
+        @stopped.make_true
+
+        stop_watcher
         stop_workers
```

Comment:
nit: unify to singularized/plurarlized terminology

Reply:
Aren't there _one_ watcher and _multiple_ workers?

Reply:
That's right. I meant for `worker` vs. `workers` (`ensure_worker_count` vs. `stop_workers` etc). I probably introduced this with my commits, but just saw it now reviewing everything.

Reply:
FWIW, "ensure_worker_count" reads more fluently to me than "ensure_workers_count". Like "word count" vs "words count", or "character count" vs "characters count".